### ISSUE TITLE
[java] CommentDefaultAccessModifier not working for classes

### ISSUE THREAD

**Issue by boris-petrov** [pmd/pmd#1793]
This is a feature-request.

The [documentation](https://pmd.github.io/pmd-6.13.0/pmd_rules_java_codestyle.html#commentdefaultaccessmodifier) specifies that this rule works only for `Method, Constructor, Field or Nested class`. Why not make it work for top-level classes?

**Comment by jsotuyod** [comment]
@boris-petrov good question. I see no real reason for the restriction, as it simply aims at documenting the choice to avoid the "this is package-private by omission". It's more unlikely to happen on top-level classes (as those are generated by the IDE almost every time), but nonetheless a valid use case.

Doing so should be trivial, just removing the `isNested()` check from:

https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/CommentDefaultAccessModifierRule.java#L90

PRs are always welcome :smile:

**Comment by boris-petrov** [comment]
@jsotuyod - thanks for the response. Actually it's not so trivial as this change leads to a NPE on `CommentDefaultAccessModifierRule.java:107` - I guess because a top-level class doesn't have a parent which is gotten on line 105. :)

My only worry is that this will work for all top-level classes. Perhaps it doesn't make sense to check top-level classes in a file where there are 2 or more because only *one* of them can be `public` and the rest *have* to be `package-private`... but I'm not sure if it is worth the hassle to make it smart and also how difficult it would be to implement.

**Comment by jsotuyod** [comment]
@boris-petrov true, then the fix would be to change that condition to something along the lines of:

```java
if ((!decl.isNested() && decl.isPackagePrivate()) || shouldReport(decl)) { 
```

**Comment by adangel** [comment]
I'm declaring this now a false-negative, since the rule misses the cases for top-level classes. Fixing this will of course create new violations for anybody, who has enabled this rule.

**Comment by matthiaskraaz** [comment]
Just been hit by your change. This is not a good change.

The code quality metric that this rule in its original state helped to improve is the scope of visibility.

Forgetting the access modifier of a potentially private or protected nested elements INCREASES the visibility to package visibility. As a result, other developers might access internal state of a class.

Completely different story with top level elements: The only possible access modifier for top level elements is public! So forgetting the access modifier of a top level elements does no harm in terms of code quality, at least not in respect to above metric.

The change calls for a lot of additional /* package */ comments without any reward in terms of code quality. Please revert or create a new rule for the changed behavior.

**Comment by matthiaskraaz** [comment]
The changed rule also warns about local classes like D in the following code:

```
public class C {
  private void f() {
    class D {
    }
  }
}
```

It is not possible to reduce the visibility of the class with any access modifier! And I think an access modifier - pseudo /* package */ or real - looks really awkward there.

**Comment by boris-petrov** [comment]
@matthiaskraaz - what you say makes sense, in the sense that one cannot forget to put a modifier and make a class *more* visible. However, explicitly stating that this is what you want is always better in my opinion. I guess this feature could be made optional...

**Comment by oowekyala** [comment]
@matthiaskraaz The rule's reporting local classes is already fixed by #1849, scheduled for the next release. 

Other than that I think it makes sense to introduce a property to silence the rule on top-level classes, like was proposed in https://github.com/pmd/pmd/pull/1798#issuecomment-489068260

I opened #1880 to track this, this issue will stay closed
